txscript/engine.go

```diff
@@ -321,6 +330,72 @@ func (vm *Engine) Step() (done bool, err error) {
 			// Set stack to be the stack from first script minus the
 			// script itself
 			vm.SetStack(vm.savedFirstStack[:len(vm.savedFirstStack)-1])
+		} else if (vm.scriptIdx == 1 && vm.witness) ||
+			(vm.witness && vm.bip16 && vm.scriptIdx == 2) { // Nested P2SH.
+			// TODO(roasbeef): check that sigscript is just a push
+			vm.scriptIdx++
+
+			witness := vm.tx.TxIn[vm.txIdx].Witness
+			if vm.witnessVersion == 0 {
+				switch len(vm.witnessProgram) {
+				case 20: // P2WKH
+					// The witness stack should consist of exactly two
+					// items: the signature, and the pubkey.
+					if len(witness) != 2 {
+						// TODO(roasbeef): actual error
+						return false, fmt.Errorf("invalid witness for p2wkh")
+					}
+
+					// Now we'll resume execution as if it were a
+					// regular p2pkh transaction.
+					pkScript, err := payToPubKeyHashScript(vm.witnessProgram)
+					if err != nil {
+						return false, err
+					}
+					pops, err := parseScript(pkScript)
+					if err != nil {
+						return false, err
+					}
+
+					// Set the stack to the provided witness
+					// stack, then append the pkScript
+					// generated above as the next script to execute.
+					vm.scripts = append(vm.scripts, pops)
+					vm.SetStack(witness)
```

Comment:
Doesn't this need to include the stack data from `vm.savedFirstStack` with the witness at the top of the stack (appended)?  Otherwise it would wipe that data out.  We need to carefully compare what Core does here because it's possible it's just wiping the stack out too, but it seems to me that it should be appended because a nested P2SH witness program could have other data pushed.